src/cart/crt_rpc.c

```diff
@@ -1537,7 +1537,7 @@ crt_req_send(crt_rpc_t *req, crt_cb_t complete_cb, void *arg)
 			/* failure already reported through complete cb */
 			if (complete_cb != NULL)
 				rc = 0;
-		} else {
+		} else if (!crt_rpc_completed(rpc_priv)) {
```

Comment:
One note to make here, we only used crt_rpc_completed() to detect bugs in the ref counting logic or elsewhere before. I don't think we should start using this function to decide ref counting, and instead should fix underlying ref count problem instead

Reply:
hmm. does this issue only happen on the issuer or the corpc or on the intermediate nodes as well that are part of corpc tree? I recall there is some special handling for 'am_root' in corpc completion and sounds like there might be ref count going wrong in that place as a result (in crt_corpc_complete()).

If you have a good reproducer are you able to instrument ADDREF/DECREF macros and add prints to track what is happening to ref counts in your failure case?

as to usage of 'crt_rpc_completed(rpc_priv)' -- it's not safe. if rpc is completed, then the access to rpc_priv is no longer valid, as the rpc_priv should be free-ed already once rpc is completed and ref count is 0; as mentioned before it was only used to previously to detect bugs in ref count logic.

Reply:
Currently, I only saw the failure happened on the coprc tree root. Not sure whether other corner cases. I traced the RPC reference count: after `crt_rpc_complete_and_unlock()` returned to `crt_corpc_complete()`, the reference was 2. And then `crt_corpc_complete()` called another RPC_DECREF(), so the reference became 1. At that time, the RPC was still valid. So it is safe to check `crt_rpc_completed()` in `crt_req_send()`.

```
int
crt_req_send(crt_rpc_t *req, crt_cb_t complete_cb, void *arg)
{
        struct crt_rpc_priv     *rpc_priv = NULL;
        bool                     locked = false;
        int                      rc = 0;
...
out:
        /* internally destroy the req when failed */
        if (rc != 0) {
                if (!rpc_priv->crp_coll) {
                        crt_rpc_complete_and_unlock(rpc_priv, rc);
                        locked = false;
                        /* failure already reported through complete cb */
                        if (complete_cb != NULL)
                                rc = 0;
                } else if (!crt_rpc_completed(rpc_priv)) {
                        RPC_DECREF(rpc_priv);
                }
        }
...
}
```

Reply:
Thank you, i understand the situation better now.

There is one more issue though with this function usage:

I just checked the crt_rpc_completed() and (since i guess it was a debug function) it not only checks for the completion but also marks it (line 485). This function would need to be split into 2 - one to set completed bit and one to query it here.

` 476 bool
 477 crt_rpc_completed(struct crt_rpc_priv *rpc_priv)
 478 {
 479         bool    rc = false;
 480
 481         D_SPIN_LOCK(&rpc_priv->crp_lock);
 482         if (rpc_priv->crp_completed) {
 483                 rc = true;
 484         } else {
 485                 rpc_priv->crp_completed = 1;
 486                 rc = false;
 487         }
 488         D_SPIN_UNLOCK(&rpc_priv->crp_lock);
 489
 490         return rc;
 491 }
`

Reply:
Hm, currently, the RPC check completion and RPC set completion are automatic operation within the `SPIN_LOCK`. If we split them into two pieces, then related callers need to hold the lock by themselves and call `check` and `set` accordingly. That seems become complex and inconvenient. The logic for `crt_rpc_completed()` was added by @liuxuezhao . I do not think it was for debug purpose. Maybe he can give more suggestion for that.

Reply:
I need a few more time to check details, I'm considering if it is better to discard the change at L1540 and with below change -

diff --git a/src/cart/crt_corpc.c b/src/cart/crt_corpc.c
index 5d7b6d79c6..8b4bfbbbed 100644
--- a/src/cart/crt_corpc.c
+++ b/src/cart/crt_corpc.c
@@ -569,10 +569,10 @@ crt_corpc_complete(struct crt_rpc_priv *rpc_priv)
                 * crt_proc_crt_bulk_t free the bulk handle again.
                 */
                rpc_priv->crp_coreq_hdr.coh_bulk_hdl = NULL;
-       }
 
-       /* correspond to addref in crt_corpc_req_hdlr */
-       RPC_DECREF(rpc_priv);
+               /* correspond to addref in crt_corpc_req_hdlr */
+               RPC_DECREF(rpc_priv);
+       }
 }
 
 static inline void

Reply:
I think it is fine to land it as it to fix the problem first, later I'll check more details to see if some details can be refined. Thx

Reply:
> I think it is fine to land it as it to fix the problem first, later I'll check more details to see if some details can be refined. Thx

@frostedcmos , how do you think that? Currently, this issue often causes CR20 to be failed during master/b26 daily tests. I prefer to make a simple fix firstly. There may be more work for related CRT logic, but we can enhance that sometime later. Thanks!

Reply:
I still dont like the fact that crt_rpc_completed() also sets completed bit on its own, as it can lead later to unintended wrong behaviors if the function is used in other places. However in the current usages it's ok to land this as a fix.

I think the better solution would be to refine completion logic for the error case when am_root==true, but we can address it later.

Reply:
I am not against the adjustment for `crt_rpc_completed`() related logic. That is not directly related with current CR test failure in DAOS-16179. So we can do that in another independent patch. As my understand, some network expert will do that, right?

BTW, please help to review the master version (https://github.com/daos-stack/daos/pull/15476) to follow our land process. Thanks!

Reply:
can consider to merge with https://github.com/daos-stack/daos/pull/15572 later